File: 101-109_small_problems/easy_9/grocery_list.rb
# Grocery List
# Write a method which takes a grocery list (array) of fruits with quantities and converts it into an array of the correct number of each fruit.

# Input: A nested array of strings and integers
# Output: One array
# Mental model: Given a nested array of strings and integers, create a new array with the strings integer number of times.

# Example:

# buy_fruit([["apples", 3], ["orange", 1], ["bananas", 2]]) ==
#   ["apples", "apples", "apples", "orange", "bananas","bananas"]

# Data Structure

# Input is an array. Output is an array.

# Algorithm

# Create new array
# Loop over input array
#   Add to new array the strings in the amount of the input.

# Return new array

# Code

def buy_fruit(list)
  new_list = []

  list.each do |item|
    amount = item.last

    amount.times do
      new_list << item.first
    end
  end

  new_list
end

p buy_fruit([["apples", 3], ["orange", 1], ["bananas", 2]]) == ["apples", "apples", "apples", "orange", "bananas","bananas"]

File: exercises/js210_small_problems/list_processing/groceryList.js
// Grocery List
// Write a function that takes a grocery list (a two-dimensional array) with each element containing a fruit and a quantity, and returns a one-dimensional array of fruits, in which each fruit appears a number of times equal to its quantity.

// Input: A nested array
// Output: A non-nested array

// What operation is this? (for the inner arrays). It's not transformation or selection.
//
// We can map the outer array, and then flatten it
//
// For the inner array, perhaps it's best to use a for loop? to loop innerArray[1] times
//
// Algorithm
// 1. Flat map outer array
// 2.   Create a new array
// 3.   Loop through inner array innerArray[1] times
// 4.    Add innerArray[0] to new array
// 5.   Return new array

function buyFruit(groceryList) {
  return groceryList.flatMap(function(fruitAndQuantity) {
    let fruit = [];

    for (let i = 0; i < fruitAndQuantity[1]; i += 1) {
      fruit.push(fruitAndQuantity[0]);
    }

    return fruit;
  });
}

//
// Example:

console.log(buyFruit([['apple', 3], ['orange', 1], ['banana', 2]]));
// returns ["apple", "apple", "apple", "orange", "banana", "banana"]
